### DISCUSSION TITLE
Java BufferedImage PNG rasterization support for MVT/pbf vector data

### DISCUSSION THREAD

**Discussion by jjzazuet** [protomaps/PMTiles#474] [General]
Hi @bdon , hope all is well!

I'm trying to implement a minimal Java server that can read and serve PMTiles content using Java only, with minimal dependencies.

I have managed to implement enough code to read a `.pmtiles` file and unpack MVT vector data from the MapBox Proto-buffer specification given some `Z/X/Y` coordinates.

- https://github.com/vaccovecrana/nugui/blob/main/src/main/java/io/vacco/nugui/PMTiles.java
- https://github.com/vaccovecrana/nugui/blob/main/src/test/java/io/vacco/nugui/PMTilesTest.java

However, I am now struggling to render the vector data into a 2d `BufferedImage` that can then be saved as a PNG raster data. I have some preliminary code, but it renders garbled images.

- https://github.com/vaccovecrana/nugui/blob/main/src/main/java/io/vacco/nugui/MVTRenderer.java

Do you know where could I find a simple MVT/PBF file with some simple lines and shapes (circles, triangles, whatever) that I could use to validate how a rendered MVT vector tile should look like?

Alternatively, do you know of any specific rasterization source code or technical notes I could reference from any other projects?

Thanks!

**Comment by bdon** [comment]
I don't think MVT to PNG conversion is really in the scope of the PMTiles project, you may be able to find test fixtures in projects like https://github.com/mapbox/vector-tile-js/tree/main/test/fixtures

For rasterization the main programs people use are Mapnik and MapLibre GL Native. Both of those may have java bindings. If you are implementing a more basic program to do that e.g. without labeling you could use just Cairo, Skia or Java equivalent instead.

**Comment by jjzazuet** [comment]
Ok great, I think I can work with those `.pbf` files.

Thanks!
